### DISCUSSION TITLE
Create pocketbase records on Android, is login using WebView the right way?

### DISCUSSION THREAD

**Discussion by xrd** [pocketbase/pocketbase#3704] [Q&A] [answered]
I have an Android app and want to save records to pocketbase. I don't want to use Dart (unless you can somehow use the dart SDK within a regular Android app; not a dart user right now). I was assuming the best way to do all this would be to create a webview which points the user to the pocketbase login flow. Then, register either a JS callback once login has completed that passes the auth token to the android app via a JS bridge between the android app and the webview. Or, perhaps I can sniff out the URL changes of the webview and grab a token from a URL during the login flow?

Is this the best way to do it inside my Android app? Or, is there a better way that I should consider?

Once I have an auth token, I can figure out the right REST commands to make a record in Java/Kotlin.

**Comment by ganigeorgiev** [comment, marked as answer]
WebViews are not recommended with OAuth2 for security reasons and some providers (eg. Google) may refuse to load their sign-in pages inside WebViews.

Instead, for Android you can use [Android Custom tabs](https://developer.chrome.com/docs/android/custom-tabs/) + deeplink as a redirect url that you can catch from your app and perform the OAuth2 code->token exchange.

In general the flow should be something similar to the ["Manual code exchange"](https://pocketbase.io/docs/authentication/#oauth2-integration) plain JS example:

0. Use a deeplink as a redirect url when creating your app.
   _With PocketBase v0.19.3 the redirect url can be anything and it is not restricted only to http/https scheme which may require [verifying the android app link](https://developer.android.com/training/app-links/verify-android-applinks) (see also https://github.com/pocketbase/pocketbase/discussions/442#discussioncomment-4160820)._

1. You need somewhere in your app to show the "Login with ..." buttons and open it via Android custom tabs.
   The sign-up urls for those button can be constructed manually or for convenience by calling the ["List auth methods" endpoint](https://pocketbase.io/docs/api-records/#list-auth-methods).
   
2. Once the user authenticates, the OAuth2 provider should trigger the deeplink redirect that your app is configured to catch (eg. via intent filter) and you can perform the OAuth2 code -> token exchange via the ["Auth with OAuth2" API endpoint](https://pocketbase.io/docs/api-records/#auth-with-oauth2) (aka. the actual PocketBase authentication).
